File: boj/14502.cpp
#include <iostream>

using namespace std;

int originAdj[8][8];
int copyAdj[8][8];
int N,M;
int minAnswer;


void makeWall(int cnt){
	if(cnt == 3){
		BFS();
		return ;
	}
	
	for(int i=0; i<N; i++){
		for(int j=0; j<M; j++){
			if(copyAdj[i][j] == 0){
				copyAdj[i][j] = 1;
				makeWall(cnt+1);
				copyAdj[i][j] = 0;
			}
		}
	}
}


void copyAdj(){
	for(int i=0; i<N; i++){
		for(int j=0; j<M; j++){
			copyAdj[i][j] = originAdj[i][j];
		}
	}
}

int main(void){
	cin >> N >> M;
	
	for(int i=0; i<N; i++){
		for(int j=0; j<M; j++){
			cin >> originAdj[i][j];
		}
	}
	
	
	
	
	
	// 1. input
	// 2. copy한 배열에 벽 3개 세우기
	// 3. copy배열을 spread
	// 4. 갯수 세고 미니멈 갱신
	// 5.  
	return 0;
}

File: boj/14502.py
from collections import deque
from itertools import combinations
import copy

def make_wall(elm,MAP):
    COPY_MAP = copy.deepcopy(MAP)
    for xy in elm:
        x=xy[0]
        y=xy[1]
        COPY_MAP[x][y]=1
    return COPY_MAP

def bfs(COPY_MAP):
    dx=[-1,0,1,0]
    dy=[0,-1,0,1]
    visited=[[False for _ in range(len(COPY_MAP[0]))] for _ in range(len(COPY_MAP))]
    q = deque()
    for i in range(len(COPY_MAP)):
        for j in range(len(COPY_MAP[0])):
            if COPY_MAP[i][j]==2:
                q.append([i,j])
                visited[i][j] = True
    
    while q:
        xy = q.popleft()
        y = xy[0]
        x = xy[1]
        for i in range(4):
            nextY = y + dy[i]
            nextX = x + dx[i]

            if nextX < 0 or nextY <0 or nextY >= len(COPY_MAP) or nextX >=len(COPY_MAP[0]):
                continue

            if COPY_MAP[nextY][nextX] == 0 and not visited[nextY][nextX] :
                COPY_MAP[nextY][nextX] = 2
                q.append([nextY,nextX])
                visited[nextY][nextX] = True

    return count_safety(COPY_MAP)


def count_safety(COPY_MAP):
    cnt=0
    for i in range(len(COPY_MAP)):
        for j in range(len(COPY_MAP[0])):
            if COPY_MAP[i][j] ==0 :
                cnt +=1
    return cnt 
        

if __name__ == "__main__":
    MAP=[]
    empty=[]
    MAX= 0
    N,M = map(int,input().split())
    for _ in range(N):
        MAP.append(list(map(int,input().split())))

    for i in range(N):
        for j in range(M):
            if MAP[i][j] == 0 :
                empty.append((i,j))

    for elm in list(combinations(empty,3)):
        MAX = max(MAX,(bfs(make_wall(elm,MAP))))
    
    print(MAX)